Add SteamVR Quck Launch feature

Iemand005 opened the pull request:
Added a toggle to choose to directly start SteamVR instead of using the steam protocol.

Benefits:
- Much faster startup and restart times
- Works offline
- No memory wasted by Steam
- Enables launching SteamVR with proxy server and https decryption active.
- Should allow using ALVR with Vive's offline Enterprise SteamVR packages.
- No more worrying about Steam throwing the "app already running" error when it's not running.

Drawbacks:
- If SteamVR isn't the default OpenXR runtime and the given path in settings is wrong it can't find SteamVR. (Could add protocol as a fallback or show a pop-up to let the user know to change their settings).
- No cloud syncing (doesn't really make a difference).


I really wanted an easier way to launch SteamVR this way, it's annoying to have to use a shortcut to the executable and having the automatic restart in ALVR launch Steam. Enabling this greatly improves the experience by removing the wait for SteamVR to launch. If there's stuff you'd like changed just let me know. I'll provide support and updates for the feature must there be bugs or issues with it.

Thank you very much!

Iemand005 commented:
Here's the difference in sartup speed of SteamVR:

Original:

https://github.com/user-attachments/assets/48134f6d-734f-4d27-b9e1-51f44c054bf8

Quick Launch:

https://github.com/user-attachments/assets/30bab1af-455a-4db0-a2e7-137d2bcff80f

zmerp commented on alvr/dashboard/src/main.rs:
```diff
@@ -62,13 +62,11 @@ fn main() {
-        .open_close_steamvr_with_dashboard
-    {
-        steamvr_launcher::LAUNCHER.lock().launch_steamvr()
+    let session = data_sources::get_read_only_local_session();
+    let steamvr_settings = &session.settings().extra.steamvr_launcher;
+
+    if steamvr_settings.open_close_steamvr_with_dashboard {
+        steamvr_launcher::LAUNCHER.lock().launch_steamvr();
```
This code seems unnecessary, it doesn't change the functionality

zmerp commented on alvr/session/src/settings.rs:
```diff
@@ -1472,6 +1472,13 @@ pub struct LoggingConfig {
 pub struct SteamvrLauncher {
     #[schema(strings(display_name = "Open and close SteamVR with dashboard"))]
     pub open_close_steamvr_with_dashboard: bool,
+    #[schema(strings(
+        display_name = "SteamVR Quick Launch",
+        help = "Launches SteamVR directly without Steam. This makes launching SteamVR significantly faster, allows SteamVR to be launched offline and avoids the \"app already running\" pop-up."
+    ))]
+    pub quick_launch_steamvr: bool,
+    #[schema(strings(display_name = "SteamVR executable path"))]
+    pub steamvr_executable_path: String,
```
I would replace these two settings with `use_steamvr_path: Switch<String>`. The help string can stay the same.

zmerp reviewed (changes requested):
Note that we are about to change these parts of the code soon, in #2993. I would say to merge this PR after that one.

Iemand005 commented:
I renamed the open with dashboard setting to open automatically to make it more recognisable at a glance what the setting does.

Here's a speed comparison of automatic opening:

With Steam:

https://github.com/user-attachments/assets/dfdc7444-61cf-4e97-aacc-143eb315c802



With quick launch:

https://github.com/user-attachments/assets/54429dd7-cdde-459a-8e21-51215ec927bc




I'd like to note some stuff about the UI:

<img width="902" height="311" alt="afbeelding" src="https://github.com/user-attachments/assets/8c4450f8-d83e-4004-8c06-e13822ef7c6e" />

The width of the text input field seems to be determined by the content length instead of the width of the actual visible control. The text input field should actually try to take up the full width available instead of being a fixed width. It also seems to shift the reset buttons on the right below the control too but not above. It should either only shift the reset button on the row where there isn't enough room, or (what I feel like is more appropriate design) it should shift both the ones above and below.

In cases where the right aligned buttons are really forced to shift outside of the window view, a horizontal scroll bar should be presented so the user can at least reach those buttons, currently it appears as if they're not there at a glance, the user would need to realize this and know they have to resize the window if this happens.

I also think it'd be a nice addition to make it so that if a button is set to automatically set it's width based on the content, set the minimum width to the height of the button and center align the contents of the button. This would make the refresh buttons square instead of the odd slightly rectangle shape they are now, which I feel would look a bit more natural.

There's also chance at overlap:

<img width="74" height="65" alt="afbeelding" src="https://github.com/user-attachments/assets/2d4b2f83-351a-4bbe-8562-b5970be241b8" />

The toggles and sliders are also not vertically centered.

Combobox is off vertically by two pixels:

<img width="198" height="64" alt="afbeelding" src="https://github.com/user-attachments/assets/9220103b-a41c-4315-bad0-e7d81fabd64b" />

The hitboxes of buttons seems to extend 3 pixels outside of the button's graphical area. It'd be a nice but not necessary addition (might require a little more computation) to have the button hitboxes of buttons with rounded corners match the shape of the buttons, so have the button only get highlighted if the cursor's tip is inside the button and not have it select the button if it's in the corner but outside the button.

There are other graphical bugs that are related to this (same things vertically like in the sidebar, the way mouse down and up is handled in combobox, touch vs mouse inputs, scrolling is jumpy, toggles can only be togged with a click and not by dragging the dot from left to right). It doesn't really affect the functionality of the application but having the UI be consistent, robust and respond to user inputs seamlessly and intuitively helps a lot to make people think less of the user interface, gives the app a premium feel and helps people focus on the settings.

I haven't looked at the UI code yet but when I have time I'll probably see if it'd be easy enough for me to make some improvements here.

Let me know if you'd like me to fix the UI a bit and make it a bit more user friendly. I'd love to work on addressing these issues.

zmerp commented:
I'm aware of the graphical bugs. We use egui, which is an immediate mode GUI library. egui is the most popular egui library for Rust, but also it moves quite fast, often rewriting the internals completely. Its theming engine is quite weak and pretty buggy, which is what you see as misalignments with the controls (using the default margins it would not produce the same misalignments). Moreover, the weirdness of the layout is caused by egui being an immediate mode gui. Immediate mode means that the library cannot produce the functionality of flexbox, centering, anything that requires knowledge of what will be drawn next (egui actually tries to be smart when using built-in controls but there are pitfalls everywhere).

Despite all of these issues, we decided to go with egui because it will allow to embed the dashboard or part of it inside the client, as it has an easy to integrate WGPU backend.

Iemand005 commented:
Awesome!

I'll probably look at the UI code later and I'll do another pull request once I've implemented some proper design tweaks. I've been using the build with the Quick Launch feature now and it works great. I think it's all ready to be integrated into ALVR for the next release.

Thank you very much!

zmerp commented on alvr/dashboard/src/steamvr_launcher/mod.rs:
```diff
@@ -107,6 +110,33 @@ fn unblock_alvr_driver_within_vrsettings(text: &str) -> Result<String> {
+        }
+    };
+
+    #[cfg(windows)]
+    return get_steamvr_root_dir
+        .join("bin")
+        .join("win64")
+        .join("vrstartup.exe")
+        .into_os_string()
+        .into_string()
+        .unwrap();
```
*OS-specific API

Iemand005 commented on alvr/session/src/settings.rs:
```diff
@@ -1468,10 +1468,24 @@ pub struct LoggingConfig {
     pub debug_groups: DebugGroupsConfig,
 }
 
+#[derive(SettingsSchema, Serialize, Deserialize, Clone)]
+pub struct SteamvrQuickLaunchConfig {
+    #[schema(strings(display_name = "SteamVR executable path"))]
+    pub steamvr_executable_path: String,
+}
```
I attempted to add a label directly but I couldn't find how. When I looked at how other settings do this they make a custom struct too. (Like if you scroll up a little to `RawEventsConfig`.)

Iemand005 commented:
Hey @The-personified-devil,

My apologies for the delay, I wasn't home for a couple days but I got back and applied the requested changes.

I hope it's a bit better, let me know if I missed some things.

Have a nice day!

The-personified-devil commented on alvr/session/src/settings.rs:
```diff
@@ -1468,10 +1468,24 @@ pub struct LoggingConfig {
     pub open_close_steamvr_with_dashboard: bool,
+    #[schema(strings(
+        display_name = "Quick launch",
+        help = "Launches SteamVR directly without Steam. This makes launching SteamVR significantly faster, allows SteamVR to be launched offline and avoids the \"app already running\" pop-up."
```
I think this reads a little too much like an advertisement of the setting, since it only mentions upsides. In the past a lot of settings have been purely descriptive, i.e. not mentioning any reasons why or why not to use it, so that's always an option to fall back to, I think most users that would actually care to use this would find it just by description.

(Also you can set Steam to be offline to launch steamvr offline. And for steamvr it's probably also good not to have multiple instances running because how would games know which one to pick if they don't just contend resources directly.)

Alternatively I think a more toned down wording would be appropriate: "This can speed up SteamVR launches, allow multiple instances at once independent of Steam, but may cause issues with OpenXR support." (Since you mentioned openxr can be an issue.)

Iemand005 replied:
It doesn't allow multiple instances of SteamVR.

SteamVR automatically makes sure there's one instance running, it wouldn't be possible to run multiple instances as there's only one correct port (27062) for the TCP socket. If this port is in use by some other program SteamVR asks you if you want to kill the process using it. SteamVR does have a bug where if it crashes (happens with the SteamVR drivers) it sometimes leaves the TCP listener open and you have to log out and back in or wait a day for it to close.

If SteamVR is already running and you run it again it doesn't create a new instance but it seems to reload some stuff as it sometimes solves issues like when SteamVR is running but not connected to ALVR, running the vrstartup again makes it connect.

I was wrong about needing SteamVR to be the default OpenXR runtime. It can find SteamVR just fine either way and the path that's placed in the field by default works even if no SteamVR location is in the registry.

Having used the feature for a while now, there's really no drawbacks as far as I can tell, everything works perfectly fine every time, it simply makes it faster.

I'll update the string for you.

The-personified-devil replied:
Well I assumed it allows multiple instances because it says it fixes the "App already running" popup. Which I assume is generated by Steam itself because direct launch fixes it. And afaik Steams detection mechanism just monitors active processes, not ports. So that seems to me like it is launching two instances (even if one is in a broken state).

If you indeed noticed it works fine w/ openxr and stuff, you don't have to mention that, I still think the string should be more "toned down" tho.

The-personified-devil commented on alvr/dashboard/src/steamvr_launcher/mod.rs:
```diff
@@ -107,6 +110,23 @@ fn unblock_alvr_driver_within_vrsettings(text: &str) -> Result<String> {
 }
 
+pub fn get_default_steamvr_executable_path() -> String {
+    let steamvr_root_dir = alvr_server_io::steamvr_root_dir()
+        .map_err(|e| error!("Couldn't find OpenVR or SteamVR files. {e}"))
+        .unwrap_or_default();
```
I think this will be fine, but I'm really not sure what the default path could be so this could end up behaving quite weird. You should use `?` instead of the `unwrap_or_default()` and having the function return an option. On the call site that would then look like `if let Some(path) = get...()  {...}`.

zmerp reviewed (changes requested):
Sorry, I forgot to publish my review a few days ago

Iemand005 commented:
No problem. I fixed it.

zmerp commented:
@Iemand005 No other comments from me. Fix the last nit from ThePersonifiedDevil and we can merge

zmerp commented on alvr/dashboard/src/steamvr_launcher/mod.rs:
```diff
@@ -156,11 +171,36 @@ impl Launcher {
-
-            #[cfg(target_os = "linux")]
-            linux_steamvr::start_steamvr();
+            if let Switch::Enabled(steamvr_path) = data_sources::get_read_only_local_session()
+                .settings()
+                .extra
+                .steamvr_launcher
+                .use_steamvr_path
+                .clone()
```
It was better not to clone, but it doesn't matter too much for this code as it's not performance sensitive.

zmerp commented on alvr/dashboard/src/steamvr_launcher/mod.rs:
```diff
@@ -156,11 +171,36 @@ impl Launcher {
+                .use_steamvr_path
+                .clone()
+            {
+                let mut steamvr_path = steamvr_path;
+                if !PathBuf::from(&steamvr_path).exists() {
+                    warn!("SteamVR executable not found at: {steamvr_path}. Trying default path.");
+
+                    steamvr_path = match get_default_steamvr_executable_path() {
```
Using mutable variables is discouraged in rust, and in this case it's clearly avoidable.

Iemand005 replied:
You're totally right, that wasn't very good code.  I was wondering how to write that to not have to duplicate the logging and command part without cloning and mutable but wasn't sure how.  I extracted it to a function now, the clone and mutable var are gone.  Should be more performant and simpler that way.  I know it's a small function that just logs and calls a command but I personally think it's a clean solution.  Let me know what you think of it!
Thank you!

zmerp reviewed (changes requested):
The code was fine as before without rewriting more stuff. You should fix the comments or just revert the code

The-personified-devil reviewed (commented):
I think this is fine for now, there's some small cleanup things I want to change but I don't want to drag this on even more with confusing review comments, so I'll just make a follow up PR in the next couple of days.
